### PR TITLE
Retry updating build info on NotFastForwardUpdateException

### PR DESCRIPTION
**Pull request by dibarbet**
See https://github.com/dotnet/buildtools/issues/2221
We hit this issue semi-regularly in our CI pipeline
https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3173501

### CODE REVIEW

**Comment by dibarbet** [comment]
@dagood Do you have a good way of running the tool locally to test?  I couldn't find anything obvious

**Comment by dagood** [comment]
Well, this is the wrong repo. 😛 This lib is now built over in dotnet/arcade: https://github.com/dotnet/arcade/tree/master/src/Microsoft.DotNet.VersionTools/lib

I generally run through changes with the MSBuild task, [`UpdatePublishedVersions`](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.VersionTools/tasks/src/UpdatePublishedVersions.cs), pointing at my own fork of dotnet/versions and my bot account (to reduce permission set).

I use Fiddler to test things like this where I need to precisely control the order of HTTP calls. (Run the task twice at the same time and manually let HTTP requests through in a way that triggers the problem.)

You could write unit tests by mocking out the `IGitHubClient` interface and adding a place to inject your own client, but I think manual testing is fine to get this merged.

(Link to Roslyn issue for reference: https://github.com/dotnet/roslyn/issues/39308.)

File: src/Microsoft.DotNet.VersionTools/Automation/GitHubVersionsRepoUpdater.cs [hunk 1]
```diff
@@ -148,21 +148,39 @@ public GitHubVersionsRepoUpdater(GitHubAuth gitHubAuth, GitHubProject project)
                     }
                     catch (HttpRequestException ex)
                     {
-                        int nextTry = i + 1;
-                        if (nextTry < MaxTries)
+                        bool retrySucceeded = await TryRetry(i, ex.Message);
```
Comment (dagood) [inline src/Microsoft.DotNet.VersionTools/Automation/GitHubVersionsRepoUpdater.cs]: This method and the variable are really confusing... after reading it all I can see what it's doing, but these names don't describe it with it as far as I can tell. I see it's reducing duplication--I suggest replacing this loop with this util that handles it to clean it all up:

https://github.com/dotnet/arcade/blob/master/src/Common/ExponentialRetry.cs

(Nit: The method is also missing `Async` suffix.)

**Comment by dibarbet** [comment]
> Well, this is the wrong repo. 😛 This lib is now built over in dotnet/arcade: https://github.com/dotnet/arcade/tree/master/src/Microsoft.DotNet.VersionTools/lib

Ahh!  I'll close this and open a new one to arcade.

**Pull request closed without merging**